File: solns/scala/4.scala
def isPalindrome(n: Int) = n.toString.reverse.toInt == n

val xs = for (
  i <- 999 to 1 by -1;
  j <- i to 1 by -1;
  if (isPalindrome(i * j))
) yield i * j

val result = xs.reduceLeft(_ max _)

println(result)
assert(result == 906609)

File: solns/csharp/4.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;


static string Reverse(string s)
{
    var cs = s.ToCharArray();
    Array.Reverse(cs);
    return new string(cs);
}

static bool IsPalindrome(long n)
{
    var s = Convert.ToString(n);
    return Reverse(s) == s;
}

var result = 0;
for (var i = 999; i > 0; i--)
{
    for (var j = i; j > 0; j--)
    {
        var x = i * j;
        if (IsPalindrome(x) && x > result)
        {
            result = x;
        }
    }
}

Console.WriteLine(result);
Trace.Assert(result == 906609);
